2-numpy/lecture_functions.py

```diff
@@ -1,52 +1,109 @@
+and making and plotting polynomials.
+"""
 import matplotlib.pyplot as plt
+import numpy as np
 
 
-def fracformat(x, pos):
-    frac = Fraction(x / np.pi)
-    if frac.numerator == 0:
-        return 0
-    elif frac.numerator == frac.denominator:
-        return r"$\pi$"
-    elif frac.numerator == 1:
-        return r"$\frac{{ \pi }}{{ {:2d} }}$".format(frac.denominator)
-    elif frac.denominator == 1:
-        return r"${:2d} \pi$".format(frac.numerator)
-    else:
-        return r"$\frac{{ {:2d} \pi }}{{ {:2d} }}$".format(
-            frac.numerator, frac.denominator
-        )
+def cart2polar(arr):
+    """
+    Convert cartesian coordinates to polar coordinates.
 
+    Parameters
+    ----------
+    arr : numpy array
+    An array of cartesian coordinates.
 
-def cart2polar(arr):
-    arr = np.array((np.linalg.norm(arr, axis=1), np.arctan2(arr[:, 1], arr[:, 0]))).T
-    return arr
+    Returns
+    -------
+    numpy array
+    An array of polar coordinates.
+    """
+    return np.array((np.linalg.norm(arr, axis=1), np.arctan2(arr[:, 1], arr[:, 0]))).T
 
 
 def cart2spherical(arr):
-    R = np.linalg.norm(arr, axis=1)
+    """
+    Convert cartesian coordinates to spherical coordinates.
+
+    Parameters
+    ----------
+    arr : numpy array
+    An array of cartesian coordinates.
+
+    Returns
+    -------
+    numpy array
+    An array of spherical coordinates.
+    """
+    r = np.linalg.norm(arr, axis=1)
     theta = np.arctan2(np.linalg.norm(arr[:, :2], axis=1), arr[:, 2])
     phi = np.arctan2(arr[:, 1], arr[:, 0])
-    arr = np.array((R, theta, phi)).T
-    return arr
+    return np.array((r, theta, phi)).T
 
 
 def polynomial_xy(npoints):
+    """
+    Generate x and y data for a polynomial.
+
+    Parameters
+    ----------
+    npoints : int
+    The number of points to generate.
+
+    Returns
+    -------
+    tuple
+    A tuple containing x and y data, both are nunpy arrays.
+    """
+    rng = np.random.default_rng()
     x = np.linspace(-5, 5, npoints)
-    y = (2 * x**2 - 2 * x + 40) + (10 * np.random.normal(size=x.shape))
+    y = (2 * x**2 - 2 * x + 40) + (10 * rng.normal(size=x.shape))
     return x, y
 
 
 def plot_polynomial(x, y, a=None, b=None, c=None):
+    """
+    Plot a polynomial.
+
+    Parameters
+    ----------
+    x : numpy array
+        The x values.
+
+    y : numpy array
+        The y values.
+
+    a : Plot a polynomial.
+
+    Parameters
+    ----------
+    x : numpy array
+        The x values.
+
+    y : numpy array
+        The y values.
+
+    a : float, optional
+        The coefficient of the x^2 term.
+
+    b : float, optional
+        The coefficient of the x term.
+
+    c : float, optional
+        The coefficient of the constant term.
+    """
     plt.figure(figsize=(4, 2))
     plt.plot(x, y, color="b", marker="o", linewidth=0, alpha=0.5)
 
-    if all([args != None for args in (a, b, c)]):
+    if all(args is not None for args in (a, b, c)):
```

Comment:
`plot_polynomial()` is used to both visualize the data and plot the solution. For the first task, it is run without `a`, `b`, and `c`. When these parameters are included, a line following them is also plotted over the data. 
I guess this was made to make it faster to solve the task since the students don't have to deal with making plots themselves. Though I'm not sure it's the best solution, it seems to work so I won't change it right now.